File: scripts/demo_msg_subscriber.py
#! /bin/python3

import rospy
from mastering_ros_demo_pkg.msg import demo_msg # Auto gen from demo.msg 

def callback(msg):
    rospy.loginfo('Recieved greeting [%s]' % msg.greeting)
    rospy.loginfo('Recieved [%s]' % msg.number)

def demo_msg_subscriber():
    # Initializing ROS node with a name of demo_msg_subscriber
    rospy.init_node('demo_msg_subscriber', anonymous=True)
    # Create a publisher object
    rospy.Subscriber('numbers_msg', demo_msg, callback)
    rospy.spin()

if __name__ == '__main__':
    try:
        demo_msg_subscriber()
    except rospy.ROSInterruptException:
        pass

File: src/demo_msg_subscriber.cpp
/*
* This code will subscriber integer values from demo_topic_publisher
*/

#include <sstream>
#include <ros/ros.h>
#include "mastering_ros_demo_pkg/demo_msg.h" /* Auto gen from demo.msg */

void number_callback(const mastering_ros_demo_pkg::demo_msg::ConstPtr& msg)
{
    ROS_INFO("Recieved  greeting [%s]",msg->greeting.c_str());
    ROS_INFO("Recieved  [%d]",msg->number);
}

int main(int argc, char **argv)
{
    /* Initializing ROS node with a name of demo_topic_subscriber */
    ros::init(argc, argv,"demo_msg_subscriber");
    /* Created a nodehandle object */
    ros::NodeHandle nh;
    /* Create a publisher object */
    ros::Subscriber number_subscriber = nh.subscribe("/demo_msg_topic",10,number_callback);

    ros::spin();
    return 0;
}
